File: 982/982.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/982.py

using System;
using System.Collections.Generic;
using System.Linq;

class Program982 {
    const double EPS_VAL = 1e-9;

    static void DoPivot(List<List<double>> tableau, List<int> basis, int row, int col) {
        double pivotVal = tableau[row][col];
        double inv = 1.0 / pivotVal;
        int nCols = tableau[0].Count;
        for (int j = 0; j < nCols; j++) tableau[row][j] *= inv;
        for (int i = 0; i < tableau.Count; i++) {
            if (i == row) continue;
            double factor = tableau[i][col];
            if (Math.Abs(factor) > EPS_VAL) {
                for (int j = 0; j < nCols; j++) tableau[i][j] -= factor * tableau[row][j];
            }
        }
        basis[row] = col;
    }

    static void SetObjective(List<List<double>> tableau, List<int> basis, List<double> c) {
        int m = basis.Count;
        int n = tableau[0].Count - 1;
        List<double> obj = new List<double>();
        for (int j = 0; j < n; j++) obj.Add(-c[j]);
        obj.Add(0.0);
        for (int i = 0; i < m; i++) {
            double cb = c[basis[i]];
            if (Math.Abs(cb) > EPS_VAL) {
                List<double> row = tableau[i];
                for (int j = 0; j <= n; j++) obj[j] += cb * row[j];
            }
        }
        if (tableau.Count == m) tableau.Add(obj);
        else tableau[m] = obj;
    }

    static bool SimplexMax(List<List<double>> tableau, List<int> basis) {
        int m = basis.Count;
        int n = tableau[0].Count - 1;
        int maxIter = 200000;
        for (int iter = 0; iter < maxIter; iter++) {
            int entering = -1;
            for (int j = 0; j < n; j++) {
                if (tableau[m][j] < -EPS_VAL) { entering = j; break; }
            }
            if (entering == -1) return true;
            double minRatio = double.PositiveInfinity;
            int leaving = -1;
            for (int i = 0; i < m; i++) {
                double a = tableau[i][entering];
                if (a > EPS_VAL) {
                    double ratio = tableau[i][tableau[i].Count - 1] / a;
                    if (ratio < minRatio - EPS_VAL) { minRatio = ratio; leaving = i; }
                }
            }
            if (leaving == -1) return false;
            DoPivot(tableau, basis, leaving, entering);
        }
        throw new Exception("Simplex did not converge");
    }

    static (List<List<double>> tableau, List<int> basis, List<int> artIndices, int nTotal)
    BuildTableau(int nVars, List<(List<double> coeffs, string sense, double b)> constraints) {
        List<List<double>> rows = new List<List<double>>();
        List<double> rhs = new List<double>();
        List<int> basis = new List<int>();
        List<int> artIndices = new List<int>();
        int nTotal = nVars;

        void AddVar() {
            foreach (var r in rows) r.Add(0.0);
            nTotal += 1;
        }

        for (int ci = 0; ci < constraints.Count; ci++) {
            List<double> coeffs = new List<double>(constraints[ci].coeffs);
            string sense = constraints[ci].sense;
            double b = constraints[ci].b;
            if (b < 0) {
                coeffs = coeffs.Select(v => -v).ToList();
                b = -b;
                if (sense == "<=") sense = ">=";
                else if (sense == ">=") sense = "<=";
            }
            List<double> row = new List<double>(coeffs);
            while (row.Count < nTotal) row.Add(0.0);
            if (sense == "<=") {
                AddVar();
                row.Add(1.0);
                basis.Add(nTotal - 1);
            } else if (sense == ">=") {
                AddVar();
                row.Add(-1.0);
                AddVar();
                row.Add(1.0);
                basis.Add(nTotal - 1);
                artIndices.Add(nTotal - 1);
            } else if (sense == "=") {
                AddVar();
                row.Add(1.0);
                basis.Add(nTotal - 1);
                artIndices.Add(nTotal - 1);
            } else {
                throw new Exception("Unknown constraint sense");
            }
            rows.Add(row);
            rhs.Add(b);
        }
        List<List<double>> tableau = rows.Select((r, i) => { var nr = new List<double>(r); nr.Add(rhs[i]); return nr; }).ToList();
        return (tableau, basis, artIndices, nTotal);
    }

    static (List<List<double>> tableau, List<int> basis, Dictionary<int,int> mapping)
    RemoveArtificial(List<List<double>> tableau, List<int> basis, List<int> artIndices) {
        HashSet<int> artSet = new HashSet<int>(artIndices);
        int m = basis.Count;
        int i = 0;
        while (i < m) {
            if (artSet.Contains(basis[i])) {
                int n = tableau[0].Count - 1;
                int pivotCol = -1;
                for (int j = 0; j < n; j++) {
                    if (artSet.Contains(j)) continue;
                    if (Math.Abs(tableau[i][j]) > EPS_VAL) { pivotCol = j; break; }
                }
                if (pivotCol != -1) {
                    DoPivot(tableau, basis, i, pivotCol);
                    i++;
                } else {
                    if (Math.Abs(tableau[i][tableau[i].Count - 1]) > EPS_VAL)
                        throw new Exception("Infeasible during artificial removal");
                    tableau.RemoveAt(i);
                    basis.RemoveAt(i);
                    m--;
                    continue;
                }
            } else { i++; }
        }
        int nFinal = tableau[0].Count - 1;
        List<int> keepCols = Enumerable.Range(0, nFinal).Where(j => !artSet.Contains(j)).ToList();
        Dictionary<int,int> mapping = new Dictionary<int,int>();
        for (int newIdx = 0; newIdx < keepCols.Count; newIdx++) mapping[keepCols[newIdx]] = newIdx;
        List<List<double>> newTableau = tableau.Select(row => {
            List<double> newRow = keepCols.Select(j => row[j]).ToList();
            newRow.Add(row[row.Count - 1]);
            return newRow;
        }).ToList();
        List<int> newBasis = basis.Select(b => mapping[b]).ToList();
        return (newTableau, newBasis, mapping);
    }

    static double SolveLP(int nVars, List<(List<double> coeffs, string sense, double b)> constraints, List<double> objective) {
        var (tableau, basis, artIndices, nTotal) = BuildTableau(nVars, constraints);
        Dictionary<int,int> mapping;
        if (artIndices.Count > 0) {
            List<double> cPhase1 = Enumerable.Repeat(0.0, nTotal).ToList();
            foreach (int j in artIndices) cPhase1[j] = -1.0;
            SetObjective(tableau, basis, cPhase1);
            if (!SimplexMax(tableau, basis)) throw new Exception("Unbounded in phase I");
            if (tableau[tableau.Count - 1][tableau[tableau.Count - 1].Count - 1] < -1e-7)
                throw new Exception("Infeasible LP");
            tableau.RemoveAt(tableau.Count - 1);
            var removeResult = RemoveArtificial(tableau, basis, artIndices);
            tableau = removeResult.tableau;
            basis = removeResult.basis;
            mapping = removeResult.mapping;
        } else {
            mapping = new Dictionary<int,int>();
            for (int k = 0; k < nTotal; k++) mapping[k] = k;
        }
        int nTotal2 = tableau[0].Count - 1;
        List<double> cPhase2 = Enumerable.Repeat(0.0, nTotal2).ToList();
        for (int j = 0; j < objective.Count; j++) {
            if (mapping.ContainsKey(j)) cPhase2[mapping[j]] = -objective[j];
        }
        SetObjective(tableau, basis, cPhase2);
        if (!SimplexMax(tableau, basis)) throw new Exception("Unbounded in phase II");
        return -tableau[tableau.Count - 1][tableau[tableau.Count - 1].Count - 1];
    }

    static List<List<int>> CartesianProduct(List<int> arr, int repeat) {
        List<List<int>> result = new List<List<int>> { new List<int>() };
        for (int r = 0; r < repeat; r++) {
            List<List<int>> next = new List<List<int>>();
            foreach (var prev in result)
                foreach (var v in arr) {
                    var newList = new List<int>(prev); newList.Add(v); next.Add(newList);
                }
            result = next;
        }
        return result;
    }

    static double BuildAndSolve(int numDice) {
        List<int> values = new List<int> { 1, 2, 3, 4, 5, 6 };
        List<List<int>> states = CartesianProduct(values, numDice);
        int numStates = states.Count;
        List<int> hideOptions = Enumerable.Range(0, numDice).ToList();

        HashSet<string> signalSet = new HashSet<string>();
        foreach (var t in states) {
            foreach (var h in hideOptions) {
                var revealed = t.Where((_, i) => i != h).OrderBy(x => x).ToList();
                signalSet.Add(string.Join(",", revealed));
            }
        }
        List<List<int>> signals = signalSet
            .Select(s => s.Split(',').Select(int.Parse).ToList())
            .OrderBy(a => string.Join(",", a.Select(x => x.ToString("D5"))))
            .ToList();
        Dictionary<string, int> signalIndex = new Dictionary<string, int>();
        for (int i = 0; i < signals.Count; i++) signalIndex[string.Join(",", signals[i])] = i;

        int numX = numStates * numDice;
        int numZ = signals.Count;
        int nVars = numX + numZ;

        int XIndex(int stateIdx, int hideIdx) => stateIdx * numDice + hideIdx;
        int ZIndex(int signalIdx) => numX + signalIdx;

        List<(List<double>, string, double)> constraints = new List<(List<double>, string, double)>();
        double pState = 1.0 / numStates;

        for (int sIdx = 0; sIdx < numStates; sIdx++) {
            List<double> coeffs = Enumerable.Repeat(0.0, nVars).ToList();
            foreach (var h in hideOptions) coeffs[XIndex(sIdx, h)] = 1.0;
            constraints.Add((coeffs, "=", pState));
        }

        foreach (var sig in signals) {
            double bVal = sig.Max();
            int sigIdx = signalIndex[string.Join(",", sig)];
            List<double> coeffsVis = Enumerable.Repeat(0.0, nVars).ToList();
            List<double> coeffsHid = Enumerable.Repeat(0.0, nVars).ToList();
            coeffsVis[ZIndex(sigIdx)] = -1.0;
            coeffsHid[ZIndex(sigIdx)] = -1.0;
            for (int sIdx = 0; sIdx < numStates; sIdx++) {
                var t = states[sIdx];
                foreach (var h in hideOptions) {
                    var revealed = t.Where((_, i) => i != h).OrderBy(x => x).ToList();
                    if (string.Join(",", revealed) == string.Join(",", sig)) {
                        coeffsVis[XIndex(sIdx, h)] += bVal;
                        double hiddenVal = t[h];
                        coeffsHid[XIndex(sIdx, h)] += hiddenVal;
                    }
                }
            }
            constraints.Add((new List<double>(coeffsVis), "<=", 0.0));
            constraints.Add((new List<double>(coeffsHid), "<=", 0.0));
        }

        List<double> objective = Enumerable.Repeat(0.0, nVars).ToList();
        for (int sigIdx = 0; sigIdx < numZ; sigIdx++) objective[ZIndex(sigIdx)] = 1.0;

        return SolveLP(nVars, constraints, objective);
    }

    static void Main(string[] args) {
        double valTwo = BuildAndSolve(2);
        double targetExact = 145.0 / 36.0;
        System.Diagnostics.Debug.Assert(Math.Abs(valTwo - targetExact) < 1e-8);
        System.Diagnostics.Debug.Assert(Math.Abs(valTwo - 4.027778) < 1e-6);
        double valThree = BuildAndSolve(3);
        Console.WriteLine(valThree.ToString("F6"));
    }
}

File: 982/982.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/982.py

val EPS_VAL = 1e-9

fun doPivot(tableau: MutableList<MutableList<Double>>, basis: MutableList<Int>, row: Int, col: Int) {
    val pivotVal = tableau[row][col]
    val inv = 1.0 / pivotVal
    val nCols = tableau[0].size
    for (j in 0 until nCols) tableau[row][j] *= inv
    for (i in tableau.indices) {
        if (i == row) continue
        val factor = tableau[i][col]
        if (Math.abs(factor) > EPS_VAL) {
            for (j in 0 until nCols) tableau[i][j] -= factor * tableau[row][j]
        }
    }
    basis[row] = col
}

fun setObjective(tableau: MutableList<MutableList<Double>>, basis: MutableList<Int>, c: List<Double>) {
    val m = basis.size
    val n = tableau[0].size - 1
    val obj = MutableList(n) { j -> -c[j] }
    obj.add(0.0)
    for (i in 0 until m) {
        val cb = c[basis[i]]
        if (Math.abs(cb) > EPS_VAL) {
            val row = tableau[i]
            for (j in 0..n) obj[j] += cb * row[j]
        }
    }
    if (tableau.size == m) tableau.add(obj)
    else tableau[m] = obj
}

fun simplexMax(tableau: MutableList<MutableList<Double>>, basis: MutableList<Int>): Boolean {
    val m = basis.size
    val n = tableau[0].size - 1
    val maxIter = 200000
    for (iter in 0 until maxIter) {
        var entering: Int? = null
        for (j in 0 until n) {
            if (tableau[m][j] < -EPS_VAL) { entering = j; break }
        }
        if (entering == null) return true
        var minRatio = Double.POSITIVE_INFINITY
        var leaving: Int? = null
        for (i in 0 until m) {
            val a = tableau[i][entering]
            if (a > EPS_VAL) {
                val ratio = tableau[i].last() / a
                if (ratio < minRatio - EPS_VAL) { minRatio = ratio; leaving = i }
            }
        }
        if (leaving == null) return false
        doPivot(tableau, basis, leaving, entering)
    }
    throw RuntimeException("Simplex did not converge")
}

data class BuildTableauResult(
    val tableau: MutableList<MutableList<Double>>,
    val basis: MutableList<Int>,
    val artIndices: MutableList<Int>,
    val nTotal: Int
)

fun buildTableau(nVars: Int, constraints: List<Triple<List<Double>, String, Double>>): BuildTableauResult {
    val rows = mutableListOf<MutableList<Double>>()
    val rhs = mutableListOf<Double>()
    val basis = mutableListOf<Int>()
    val artIndices = mutableListOf<Int>()
    var nTotal = nVars

    fun addVar() {
        for (r in rows) r.add(0.0)
        nTotal += 1
    }

    for (ci in constraints.indices) {
        var coeffs = constraints[ci].first.toMutableList()
        var sense = constraints[ci].second
        var b = constraints[ci].third
        if (b < 0) {
            coeffs = coeffs.map { -it }.toMutableList()
            b = -b
            if (sense == "<=") sense = ">=" else if (sense == ">=") sense = "<="
        }
        val row = coeffs.toMutableList()
        while (row.size < nTotal) row.add(0.0)
        when (sense) {
            "<=" -> {
                addVar()
                row.add(1.0)
                basis.add(nTotal - 1)
            }
            ">=" -> {
                addVar()
                row.add(-1.0)
                addVar()
                row.add(1.0)
                basis.add(nTotal - 1)
                artIndices.add(nTotal - 1)
            }
            "=" -> {
                addVar()
                row.add(1.0)
                basis.add(nTotal - 1)
                artIndices.add(nTotal - 1)
            }
            else -> throw IllegalArgumentException("Unknown constraint sense")
        }
        rows.add(row)
        rhs.add(b)
    }
    val tableau = rows.mapIndexed { i, r -> (r + mutableListOf(rhs[i])).toMutableList() }.toMutableList()
    return BuildTableauResult(tableau, basis, artIndices, nTotal)
}

data class RemoveArtResult(
    val tableau: MutableList<MutableList<Double>>,
    val basis: MutableList<Int>,
    val mapping: Map<Int, Int>
)

fun removeArtificial(tableau: MutableList<MutableList<Double>>, basis: MutableList<Int>, artIndices: List<Int>): RemoveArtResult {
    val artSet = artIndices.toHashSet()
    var m = basis.size
    var i = 0
    while (i < m) {
        if (basis[i] in artSet) {
            val n = tableau[0].size - 1
            var pivotCol: Int? = null
            for (j in 0 until n) {
                if (j in artSet) continue
                if (Math.abs(tableau[i][j]) > EPS_VAL) { pivotCol = j; break }
            }
            if (pivotCol != null) {
                doPivot(tableau, basis, i, pivotCol)
                i++
            } else {
                if (Math.abs(tableau[i].last()) > EPS_VAL)
                    throw RuntimeException("Infeasible during artificial removal")
                tableau.removeAt(i)
                basis.removeAt(i)
                m--
                continue
            }
        } else { i++ }
    }
    val n = tableau[0].size - 1
    val keepCols = (0 until n).filter { it !in artSet }
    val mapping = keepCols.mapIndexed { newIdx, old -> old to newIdx }.toMap()
    val newTableau = tableau.map { row ->
        val newRow = keepCols.map { j -> row[j] }.toMutableList()
        newRow.add(row.last())
        newRow
    }.toMutableList()
    val newBasis = basis.map { mapping[it]!! }.toMutableList()
    return RemoveArtResult(newTableau, newBasis, mapping)
}

fun solveLP(nVars: Int, constraints: List<Triple<List<Double>, String, Double>>, objective: List<Double>): Double {
    val buildResult = buildTableau(nVars, constraints)
    var tableau = buildResult.tableau
    var basis = buildResult.basis
    val artIndices = buildResult.artIndices
    var nTotal = buildResult.nTotal
    var mapping: Map<Int, Int>
    if (artIndices.isNotEmpty()) {
        val cPhase1 = MutableList(nTotal) { 0.0 }
        for (j in artIndices) cPhase1[j] = -1.0
        setObjective(tableau, basis, cPhase1)
        if (!simplexMax(tableau, basis)) throw RuntimeException("Unbounded in phase I")
        if (tableau.last().last() < -1e-7) throw RuntimeException("Infeasible LP")
        tableau.removeAt(tableau.size - 1)
        val removeResult = removeArtificial(tableau, basis, artIndices)
        tableau = removeResult.tableau
        basis = removeResult.basis
        mapping = removeResult.mapping
    } else {
        mapping = (0 until nTotal).associateWith { it }
    }
    val nTotal2 = tableau[0].size - 1
    val cPhase2 = MutableList(nTotal2) { 0.0 }
    for (j in objective.indices) {
        if (mapping.containsKey(j)) cPhase2[mapping[j]!!] = -objective[j]
    }
    setObjective(tableau, basis, cPhase2)
    if (!simplexMax(tableau, basis)) throw RuntimeException("Unbounded in phase II")
    return -tableau.last().last()
}

fun cartesianProduct(arr: List<Int>, repeat: Int): List<List<Int>> {
    var result = listOf(listOf<Int>())
    for (r in 0 until repeat) {
        val next = mutableListOf<List<Int>>()
        for (prev in result) for (v in arr) next.add(prev + v)
        result = next
    }
    return result
}

fun buildAndSolve(numDice: Int): Double {
    val values = listOf(1, 2, 3, 4, 5, 6)
    val states = cartesianProduct(values, numDice)
    val numStates = states.size
    val hideOptions = (0 until numDice).toList()

    val signalSet = mutableSetOf<List<Int>>()
    for (t in states) {
        for (h in hideOptions) {
            val revealed = t.filterIndexed { i, _ -> i != h }.sorted()
            signalSet.add(revealed)
        }
    }
    val signals = signalSet.sortedWith(Comparator { a, b ->
        for (i in 0 until minOf(a.size, b.size)) {
            val cmp = a[i].compareTo(b[i])
            if (cmp != 0) return@Comparator cmp
        }
        a.size.compareTo(b.size)
    })
    val signalIndex = signals.mapIndexed { i, s -> s to i }.toMap()

    val numX = numStates * numDice
    val numZ = signals.size
    val nVars = numX + numZ

    fun xIndex(stateIdx: Int, hideIdx: Int) = stateIdx * numDice + hideIdx
    fun zIndex(signalIdx: Int) = numX + signalIdx

    val constraints = mutableListOf<Triple<List<Double>, String, Double>>()
    val pState = 1.0 / numStates

    for (sIdx in 0 until numStates) {
        val coeffs = MutableList(nVars) { 0.0 }
        for (h in hideOptions) coeffs[xIndex(sIdx, h)] = 1.0
        constraints.add(Triple(coeffs, "=", pState))
    }

    for (sig in signals) {
        val bVal = sig.max()!!.toDouble()
        val sigIdx = signalIndex[sig]!!
        val coeffsVis = MutableList(nVars) { 0.0 }
        val coeffsHid = MutableList(nVars) { 0.0 }
        coeffsVis[zIndex(sigIdx)] = -1.0
        coeffsHid[zIndex(sigIdx)] = -1.0
        for (sIdx in states.indices) {
            val t = states[sIdx]
            for (h in hideOptions) {
                val revealed = t.filterIndexed { i, _ -> i != h }.sorted()
                if (revealed == sig) {
                    coeffsVis[xIndex(sIdx, h)] += bVal
                    val hiddenVal = t[h].toDouble()
                    coeffsHid[xIndex(sIdx, h)] += hiddenVal
                }
            }
        }
        constraints.add(Triple(coeffsVis.toList(), "<=", 0.0))
        constraints.add(Triple(coeffsHid.toList(), "<=", 0.0))
    }

    val objective = MutableList(nVars) { 0.0 }
    for (sigIdx in 0 until numZ) objective[zIndex(sigIdx)] = 1.0

    return solveLP(nVars, constraints, objective)
}

fun main() {
    val valTwo = buildAndSolve(2)
    val targetExact = 145.0 / 36.0
    check(Math.abs(valTwo - targetExact) < 1e-8) { "valTwo check 1 failed" }
    check(Math.abs(valTwo - 4.027778) < 1e-6) { "valTwo check 2 failed" }
    val valThree = buildAndSolve(3)
    println("%.6f".format(valThree))
}
